### ISSUE TITLE
Tokenization related problem

### ISSUE THREAD

**Issue by samin9796** [facebookresearch/wav2letter#793]
I used sentence piece toolkit to do tokenization. After encoding, all the spaces were replaced with this special token "▁".  Although while training I set the word separator value to '▁' , I get 99% WER and 10% CER. I read the predictions: all the words were correctly predicted but there was issues with the spaces. I found that in the test.lst.viterbi.ref file, there are spaces. So, I changed the test.lst file replacing the spaces with "▁".  However, now in the test.lst.viterbi.ref file, all the samples are unknown.

This is the head of my test.lst file (After replacing the spaces):
 `test1 /data/ahnaf/speech_dataset/our_dataset/part1/wav1/a0/TNF01_oishi_p_1_S_1-01.wav 12390.0 ▁০১কথোপকথন▁তুই▁নাকি▁আরব▁দেশে▁যাচ্ছিস▁বাদশা▁কিছু▁বলে▁না▁বোন▁তার▁হাত▁ধরে▁আবার▁বলে▁সত্যি▁বাদশাকে▁ভাত▁বেড়ে▁দিয়ে▁বলে▁তলে▁তলে▁কখন▁তুই▁ঠিক▁করলি
test2 /data/ahnaf/speech_dataset/our_dataset/part1/wav1/a1/TNF01_OISHI_p_2_10_S_1-09.wav 13470.0 ▁গত▁মঙ্গলবার▁বিকেলে▁সোনালি▁রোদ্দুরে▁একঝাঁক▁মেয়ের▁একাগ্র▁অনুশীলন▁দেখেই▁বোঝা▁যায়▁এটাই▁সেই▁মাঠ▁যেখানে▁ফুটবলার▁হওয়ার▁প্রথম▁দীক্ষা▁পেয়েছে▁তহুরা▁মারিয়া▁শামসুন্নাহার▁মার্জিয়ারা

test3 /data/ahnaf/speech_dataset/our_dataset/part1/wav1/a1/TNF01_OISHI_p_2_10_S_1-30.wav 6560.0 ▁নাম▁প্রকাশে▁অনিচ্ছুক▁এক▁শিক্ষক▁বললেন▁সবাই▁সুনামের▁ভাগীদার▁হতে▁চায়▁এ▁থেকেই▁তৈরি▁হয়েছে▁বিভেদ▁দ্বন্দ্ব
test4 /data/ahnaf/speech_dataset/our_dataset/part1/wav1/a1/TNF01_OISHI_p_2_10_S_1-31.wav 10880.0 ▁এই▁দ্বন্দ্বে▁যে▁ফুটবলাররা▁ক্ষতিগ্রস্ত▁হচ্ছে▁পিছিয়ে▁পড়ছে▁কলসিন্দুরের▁ফুটবল▁সেটির▁প্রমাণ▁গত▁দুই▁বছরে▁বঙ্গমাতা▁ফজিলাতুন্নেছা▁গোল্ড▁কাপে▁তাদের▁ফলাফল▁ভালো▁ছিল▁না`

This is the test.lst.viterbi.ref file:
`<unk> (test6080)
<unk> (test5066)
<unk> (test2505)
<unk> (test5276)`

From the terminal (test.lst file has spaces in this case):
|T|:▁ক ালো ▁র ঙ ের ▁সাথে ▁ম ের ুন স হ ▁বিভিন্ন ▁র ঙ ের ▁ড িজ াই নের ▁সাথে ▁ ঐ শ ্ব র িয়ার ▁স াজ ▁স জ ্জ া টি ▁স াজ ানো ▁ছিল ▁ক ান ▁রা ণ ী ▁সহ জ েই ▁রে ড ▁কার ্প ে টে
|P|:▁ক ালো ▁র ঙ ের ▁সাথে ▁ব ের ুন স হ ▁বিভিন্ন ▁র ঙ ের ▁ড িজ াই নের ▁সাথে ▁ ঐ শ ্ব র িয়ার ▁স াজ ▁স জ ্জ া টি ▁স াজ ানো ▁ছিল ▁ক ান ▁রা ণ ী ▁সহ জে ▁রে ড ▁কার ্প ে টে
It shows 100% WER but I see most of the words are correctly predicted.

**Comment by tlikhomanenko** [comment]
@samin9796 Did you set also flag `--usewordpiece=true`?

**Comment by samin9796** [comment]
@tlikhomanenko  For training? No, I didn't.

**Comment by tlikhomanenko** [comment]
when using wordpieces you need to set `wordseparator` as well as `usewordpiece=true`. Could you try to run test.cpp with `usewordpiece=true` at first, and maybe retrain with it too?

**Comment by samin9796** [comment]
@tlikhomanenko Yes, I just tested with usewordpiece=true and got much better WER. I will retrain the model though. 
Well, I wanted to see how it performs with OOV words. Could you please tell me how to train a lexicon free AM? Any changes in the flags?

**Comment by tlikhomanenko** [comment]
The thing that these flags just specify how to process `_`. Your lexicon defines how to map words. So just prepare all words mapping into tokens sequence and in case of word pieces we only need these flags I pointed, feel free to chekc our configs from Librispeech sota models https://github.com/facebookresearch/wav2letter/tree/v0.2/recipes/models/sota/2019

**Comment by tlikhomanenko** [comment]
Closing issue for now, feel free to reopen if you will have problems after retraining.

**Comment by samin9796** [comment]
@tlikhomanenko  
Hi! I would like to ask you a quick question- when you trained with 10,000 word pieces, what was the input channel of the last linear layer? I mean do I need to have more input channels than the output channels (number of output tokens) in the last FC layer?

**Comment by tlikhomanenko** [comment]
We are using NLABEL to have the correct linear layer to map embedding into the used number of tokens, like here https://github.com/facebookresearch/wav2letter/blob/v0.2/recipes/models/sota/2019/am_arch/am_tds_ctc.arch#L38 (if I correctly understood your question). You can see that we had 1140 which we map to the 10k. Often you have bottleneck layer, like in vision and NPL, where you map say 512 vector into nTokens, which can be several hundred thousands.

**Comment by samin9796** [comment]
@tlikhomanenko 
I retrained the model using word pieces. While training, I see 15% WER on validation set but when I run test.cpp file using the same validation set, I get 23% WER. Why is WER increasing so much while testing?

**Comment by tlikhomanenko** [comment]
Can you point how are you running Test.cpp? (you should set `--uselexicon=false` at least)

**Comment by samin9796** [comment]
I didn't use that flag. Now getting better WER. Thank you!

**Comment by samin9796** [comment]
@tlikhomanenko 
I wanted to run Test.cpp for different test sets. But I am getting this error.
`what():  Unknown entry in dictionary: '▁১৮'`

Is there any way to tackle this error apart from adding new tokens to the token set?

**Comment by samin9796** [comment]
@tlikhomanenko 
I trained another model with 3000 tokens. Now I get more than 100% WER while testing but during training time, I get ~14% WER.
This is test.cfg:

> --am=/.../002_model_validation_updated.lst.bin
--datadir=/data/ahnaf/wav2letter/dataset_prep/
--test=validation_updated.lst
--tokensdir=/data/ahnaf/wav2letter/dataset_prep/tokenization/
--tokens=v_3000_token.txt
--lexicon=/data/..../tokenization/v_3000_lexicon.txt
--maxload=-1
--show
--uselexicon=false
--wordseparator=▁
--usewordpiece=true

**Comment by tlikhomanenko** [comment]
Can you send you training command + flags too?

**Comment by samin9796** [comment]
I can see that all the letters and words are wrongly predicted.

Training command:
`mpirun --allow-run-as-root -n 7 ./Train train --flagsfile=/...../train.cfg --logtostderr=1 --reportiters=0`

Training flags:

> --runname=....
--rundir=/..../tokenization/
--datadir=/..../dataset_prep/
--tokensdir=/..../tokenization/
--train=train_updated.lst
--valid=validation_updated.lst
--lexicon=/..../v_1000_lexicon.txt
--input=wav
--tokens=v_1000_token.txt
--lr=0.1
--maxgradnorm=0.2
--momentum=0.8
--batchsize=4
--iter=921600
--lrcrit=0.006
--archdir=/...../dataset_prep/
--arch=network.arch
--enable_distributed=true
--criterion=ctc
--mfcc=true
--framesizems=30
--framestridems=20
--mfcccoeffs=18
--wordseparator=▁
--usewordpiece=true

**Comment by tlikhomanenko** [comment]
I see your tokens in test cpp is  --tokens=v_3000_token.txt while in train it is  --tokens=v_1000_token.txt

**Comment by samin9796** [comment]
Well, it is not a problem. I copied from another train.cfg file (all the flags are exactly same). I double-checked that it is not a problem.

**Comment by samin9796** [comment]
Just to be clear, I just tested two different models with 1k tokens and 2k tokens respectively. There are no issues with these models. But with the model with 3k tokens, I am facing this problem.

**Comment by tlikhomanenko** [comment]
Could you try to run your 3k tokens model with test.cpp only with these flags (the remain flags will be loaded from the model itself)?
```
--am=/.../002_model_validation_updated.lst.bin
--datadir=/data/ahnaf/wav2letter/dataset_prep/
--test=validation_updated.lst
--maxload=-1
--show
--uselexicon=false
```

**Comment by samin9796** [comment]
I have tried it now and got the following error:
`what():  Unknown index in dictionary '2999'`

**Comment by samin9796** [comment]
@tlikhomanenko 
I somehow managed to get it right. At first, I didn't mention the path of the lexicon file for running test.cpp. I was getting unknown index in dict error. So, I mistakenly added that token to the token set.  Now that I removed the token from the token set and specified the lexicon file while testing, things are going right.

**Comment by tlikhomanenko** [comment]
Glad to hear this!
